core-api/src/main/java/com/optimizely/ab/notification/NotificationCenter.java

```diff
@@ -181,75 +139,49 @@ public void onTrack(@Nonnull String eventKey, @Nonnull String userId, @Nonnull M
+            case Activate:
+                return addActivateNotificationListener((ActivateNotificationListener) notificationListener);
+            default:
+                throw new OptimizelyRuntimeException("Unsupported notificationType");
         }
-        int id = this.notificationListenerID.incrementAndGet();
-        notificationsListeners.get(notificationType).add(new NotificationHolder(id, notificationListener));
-        logger.info("Notification listener {} was added with id {}", notificationListener.toString(), id);
-        return id;
     }
 
     /**
-     * Remove the notification listener based on the notificationId passed back from addNotificationListener.
+     * Remove the notification listener based on the notificationId passed back from addDecisionNotificationHandler.
      *
      * @param notificationID the id passed back from add notification.
      * @return true if removed otherwise false (if the notification is already registered, it returns false).
      */
     public boolean removeNotificationListener(int notificationID) {
-
-        for (List<NotificationHolder> notificationHolders : notificationsListeners.values()) {
-            if (removeNotificationListener(notificationID, notificationHolders)) {
-                return true;
-            }
-        }
-
-        if (removeNotificationListener(notificationID, decisionListenerHolder)) {
-            return true;
-        }
-
-        logger.warn("Notification listener with id {} not found", notificationID);
-
-        return false;
-    }
-
-    /**
-     * Helper method to iterate find NotificationHolder in an List identified by the notificationId
-     *
-     * @param notificationID the id passed back from add notification.
-     * @param notificationHolderList list from which to remove notification listener.
-     * @return true if removed otherwise false
-     */
-    private boolean removeNotificationListener(int notificationID, List<NotificationHolder> notificationHolderList) {
-        for (NotificationHolder holder : notificationHolderList) {
-            if (holder.notificationId == notificationID) {
-                notificationHolderList.remove(holder);
+        for (NotificationManager<?> manager : notifierMap.values()) {
+            if (manager.remove(notificationID)) {
                 logger.info("Notification listener removed {}", notificationID);
                 return true;
             }
         }
 
+        logger.warn("Notification listener with id {} not found", notificationID);
         return false;
     }
 
     /**
      * Clear out all the notification listeners.
      */
     public void clearAllNotificationListeners() {
-        decisionListenerHolder.clear();
-        for (NotificationType type : NotificationType.values()) {
-            clearNotificationListeners(type);
+        for (NotificationManager<?> manager : notifierMap.values()) {
```

Comment:
Minor: just clearing the entire `notifierMap` would be more consistent with the current scheme of lazy population used with `putIfAbsent`

Reply:
Update: We're now creating a unmodifiable map containing all possible NotificationManager types, so clearing the individual managers is required.